cmd/syncthing/monitor.go

```diff
@@ -317,6 +326,87 @@ func restartMonitorWindows(args []string) error {
 	return cmd.Start()
 }
 
+// rotatedFile keeps a set of rotating logs. There will be the base file plus up
+// to maxFiles rotated ones, each ~ maxSize bytes large.
+type rotatedFile struct {
+	name            string
+	open            openFn
+	maxSize         int64 // bytes
+	maxFiles        int
+	minStatInterval time.Duration // wait at least this long between stat() calls
+	current         io.WriteCloser
+	lastStat        time.Time
+}
+
+type openFn func(name string) (io.WriteCloser, error)
+
+func newRotatedFile(name string, open openFn, maxSize int64, maxFiles int, minStatInterval time.Duration) *rotatedFile {
+	return &rotatedFile{
+		name:            name,
+		open:            open,
+		maxSize:         maxSize,
+		maxFiles:        maxFiles,
+		minStatInterval: minStatInterval,
+	}
+}
+
+func (r *rotatedFile) Write(bs []byte) (int, error) {
+	// Do the rotate check at most once a minStatInterval once any given file is
+	// open. Limits the number of stat() calls when there's a lot of logging,
+	// and limits the number of warnings if we're stuck unable to rotate for
+	// whatever reason.
+	if r.current != nil && time.Since(r.lastStat) >= r.minStatInterval {
+		info, err := os.Lstat(r.name)
+		if err != nil && !os.IsNotExist(err) {
+			// N.B. we cant call the logger from inside the log writer or we'll
+			// cause a paradox and the universe will implode.
+			fmt.Println("LOG: Rotating log:", err)
+		}
+
+		// If the current size of the log plus what we're going to write would
+		// exceed the max, close this log file and we will open another one
+		// later.
+		if err == nil && info.Size()+int64(len(bs)) > r.maxSize {
+			r.current.Close()
+			r.current = nil
+		}
+		r.lastStat = time.Now()
+	}
+
+	// Rotate old files out of the way and create a new one.
+	if r.current == nil {
+		r.rotate()
+		fd, err := r.open(r.name)
+		if err != nil {
+			return 0, err
+		}
+		r.current = fd
+		r.lastStat = time.Now() // we know it's zero bytes right now
+	}
+
+	return r.current.Write(bs)
+}
+
+func (r *rotatedFile) rotate() {
+	// The files are named "name", "name.0", "name.1", ...
+	// "name.(r.maxFiles-1)". Increase the numbers on the
+	// suffixed ones.
+	for i := r.maxFiles - 1; i > 0; i-- {
+		from := fmt.Sprintf("%s.%d", r.name, i-1)
+		to := fmt.Sprintf("%s.%d", r.name, i)
+		err := os.Rename(from, to)
```

Comment:
I am not sure if "rename on top" works on all platforms tho?

Reply:
Test pass on Windows, which is the only one that I would expect trouble from 🤷‍♂